### PR TITLE
SB can be configued with CRD when deployed with injector

### PR DESCRIPTION
**Pull request by hughsaunders**
CRD = Custom Resource Definition

This commit expands the functionality of the
"sidecar-injector.cyberark.com/secretlessConfig" annotation.

This annotation now has two forms:
1. The name of a configmap that contains a configfile
2. configmgr#configspec.

Examples of option 2:
a. k8s/crd#crdname
b. configfile/configmapname

Note that b. is different from the -config-mgr cli arg, as that
requires configfile#fspath.

This commit copies the service account token volume vmount
from an existing container in the pod to the injected pod.
This ensures that injected containers can access the k8s api.
This is a fragile hack and should be removed once KEP-36 is
implemented. That solves the plugin ordering problem at the
k8s api level, and means that sidecars added by mutating
webhooks will have service acocunt token volume mounts
added in the same way as explicitly defined containers.

Other changes:
* Reformat with gofmt
* Add Jenkinsfile and integration test
* Add CRD to helm chart and manual deployment process. Secretless
  broker will create the CRD on startup, however this is too late
  in the sidecar injection case. This is required
  when sidecar injection is used as secretless config objects need
  to be created before the secretless broker is started.

Related: #1

### CODE REVIEW

File: pkg/inject/server.go [hunk 1]
```diff
@@ -88,61 +87,66 @@ func (whsvr *WebhookServer) mutate(ar *v1beta1.AdmissionReview) *v1beta1.Admissi
...
-                Message: errMsg,
-            },
-        }
-    }
+	ag := &annotationGetter{metadata: &pod.ObjectMeta}
+	injectType := ag.getAnnotation(annotationInjectTypeKey)
+	containerMode := ag.getAnnotation(annotationContainerModeKey)
+	containerName := ag.getAnnotation(annotationContainerNameKey)
+	if ag.err != nil {
+		return failWithResponse(fmt.Sprintf("Mutation failed for pod %s, in namespace %s, due to %s", pod.Name, req.Namespace, ag.err.Error()))
```
Comment (doodlesbykumbi) [inline pkg/inject/server.go]: This is a nice pattern :sparkles:. Errors are indeed values!

File: pkg/inject/server.go [hunk 2]
```diff
@@ -88,61 +87,66 @@ func (whsvr *WebhookServer) mutate(ar *v1beta1.AdmissionReview) *v1beta1.Admissi
...
+	}
+
+	var sidecarConfig *PatchConfig
+	switch injectType {
+	case "secretless":
```
Comment (doodlesbykumbi) [inline pkg/inject/server.go]: This isn't related to the substance of the PR. These switch cases are quite busy. Seems like a good time to move them into functions.

File: pkg/inject/secretless.go [hunk 3]
```diff
@@ -1,55 +1,120 @@
...
+	// and calls to that in server.go
+	volumeMounts := []corev1.VolumeMount{
+		{
+			Name:      ServiceAccountTokenVolumeName,
+			ReadOnly:  true,
+			MountPath: "/var/run/secrets/kubernetes.io/serviceaccount",
+		},
+	}
+
+	// Three options for secretlessConfig
```
Comment (doodlesbykumbi) [inline pkg/inject/secretless.go]: I like that this maintains backwards compatibility :)... but feel free to rip apart the contract and start from scratch if some other interface makes more sense. `Naming and functionality are still subject to breaking changes.`

File: pkg/inject/secretless.go [hunk 4]
```diff
@@ -1,55 +1,120 @@
...
+	// do it here because the serviceaccount plugin which adds the SATVM is
+	// executed before this plugin injects the sidecar. KEP-36 will solve this
+	// ordering problem by re-running plugins after mutation. Then if we add a
+	// container, the serviceaccount plugin will reprocess the manifest and add the
+	// appropriate mount
+	//
+	// ** Remove me when KEP-36 lands **
+	// also remove common.getServiceAccountTokenVolumeName
+	// and calls to that in server.go
+	volumeMounts := []corev1.VolumeMount{
```
Comment (doodlesbykumbi) [inline pkg/inject/secretless.go]: Is it always the case that the SATVM doesn't happen for the injected sidecar ? I'm wondering if there are cases where the SATVM works as expected and what the impact of doubly adding SATVM would be (since I haven't spotted a check for it). Are volume mounts idempotent ?

File: pkg/inject/server.go (hunk 1 above)
Comment (hughsaunders) [reply pkg/inject/server.go]: Though it had a side effect of making all annotations required, which wasn't the previous behaviour or in line with the docs, so I've reverted it.

File: pkg/inject/secretless.go (hunk 4 above)
Comment (hughsaunders) [reply pkg/inject/secretless.go]: I'm pretty sure with the current implementation there aren't edge cases where the mount will exist as plugin ordering is hard coded. We'll need to revise when the upstream issue is fixed though.

File: tests/echoserver.yaml.sh [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+subjects:
+- kind: ServiceAccount
+  name: secretless-crd
+  namespace: $(current_namespace)
+roleRef:
+  kind: ClusterRole
+  name: secretless-crd
+  apiGroup: rbac.authorization.k8s.io
+
+## This should go into the sci helm chart & the manual install docs
```
Comment (hughsaunders) [inline tests/echoserver.yaml.sh]: TODO

File: pkg/inject/server.go (hunk 2 above)
Comment (hughsaunders) [reply pkg/inject/server.go]: Good point, but left as-is for now, this patch has been in progress too long already :)

File: pkg/inject/secretless.go (hunk 3 above)
Comment (hughsaunders) [reply pkg/inject/secretless.go]: Good point, but left as-is for now, this patch has been in progress too long already :)

File: pkg/inject/server.go [hunk 6]
```diff
@@ -88,28 +87,35 @@ func (whsvr *WebhookServer) mutate(ar *v1beta1.AdmissionReview) *v1beta1.Admissi
 		}
 	}
 
-    injectType, err := getAnnotation(&pod.ObjectMeta, annotationInjectTypeKey)
+	injectType, err := getAnnotation(&pod.ObjectMeta, annotationInjectTypeKey)
     containerMode, err := getAnnotation(&pod.ObjectMeta, annotationContainerModeKey)
```
Comment (doodlesbykumbi) [inline pkg/inject/server.go]: this and the next line have some funky indentation

File: tests/secrets.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+GCLOUD_CLUSTER_NAME: !var ci/google-container-engine-testbed/gcloud-cluster-name
+GCLOUD_ZONE: !var ci/google-container-engine-testbed/gcloud-zone
+GCLOUD_PROJECT_NAME: !var ci/google-container-engine-testbed/gcloud-project-name
+GCLOUD_SERVICE_KEY: !var:file ci/google-container-engine-testbed/gcloud-service-key
+
+DOCKER_REGISTRY_URL: us.gcr.io
+DOCKER_REGISTRY_PATH: us.gcr.io/conjur-gke-dev
```
Comment (doodlesbykumbi) [inline tests/secrets.yml]: missing newline at EOF

File: tests/tests_within_docker [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+
+    kubectl delete --wait=true --ignore-not-found=true service secretless-broker secretless-broker-external &
+
+    kubectl delete --wait=true --ignore-not-found=true clusterrole secretless-crd &
+    kubectl delete --wait=true --ignore-not-found=true rolebinding secretless-crd &
+    kubectl delete --wait=true --ignore-not-found=true clusterrolebinding secretless-crd &
+    kubectl delete --wait=true --ignore-not-found=true serviceaccount secretless-crd &
+
+    kubectl delete --wait=true --ignore-not-found=true deployment sb-sci-echoserver &
+    kubectl delete --wait=true --ignore-not-found=true pods -l app=sb-sci-echoserver &
```
Comment (doodlesbykumbi) [inline tests/tests_within_docker]: shouldn't we have a wait at the end of all these background execs ?

File: tests/tests_within_docker [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            && break
+        sleep 1
+    done
+
+    echo ">>--- Found CSR, aproving"
+    kubectl \
+        -n "${I_NAMESPACE}" \
+        certificate approve "${cert}"
+
+    #secretless-sci-605bbd0e66e-cyberark-sidecar-injector.injectors-test-sidecar-605bbd0e66e
```
Comment (doodlesbykumbi) [inline tests/tests_within_docker]: left over debug comments ?

File: tests/tests_within_docker [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+    echo "UTID: ${UNIQUE_TEST_ID}"
+
+    # deploy injector
+    # This creates deploys a sidecard injector pod. This listens for deployment
+    # creations and modifies them in flight to add a secretless-broker container,
+    # if the appropriate annotations are found.
+    announce "Deploying Sidecar Injector"
+    deploy_injector
+
+    # deploy secretless deps and test app (echo server)
```
Comment (doodlesbykumbi) [inline tests/tests_within_docker]: this is amazing. thank you!!!

File: tests/tests_within_docker [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+    echo ">>--- Sidecar Injector pod running"
+    echo ">>--- Sidecar Injector deployment complete."
+}
+
+function main() {
+    echo "using SECRETLESS_CRD_SUFFIX=${SECRETLESS_CRD_SUFFIX}"
+    echo "UTID: ${UNIQUE_TEST_ID}"
+
+    # deploy injector
+    # This creates deploys a sidecard injector pod. This listens for deployment
```
Comment (doodlesbykumbi) [inline tests/tests_within_docker]:
```suggestion
    # This creates deploys a sidecar injector pod. This listens for deployment
```

File: pkg/inject/secretless.go (hunk 4 above)
Comment (doodlesbykumbi) [reply pkg/inject/secretless.go]: I'm thinking 
1. v_now has the bug
2. v_later will not have the bug
3. if the code conditionally adds the service account volume mount then it will be compatible with both v_now and v_later

File: pkg/inject/server.go (hunk 6 above)
Comment (hughsaunders) [reply pkg/inject/server.go]: Good spot, fixed.

File: tests/secrets.yml (hunk 7 above)
Comment (hughsaunders) [reply tests/secrets.yml]: Fixed

File: tests/tests_within_docker (hunk 8 above)
Comment (hughsaunders) [reply tests/tests_within_docker]: Yes! Done.

File: tests/tests_within_docker (hunk 11 above)
Comment (hughsaunders) [reply tests/tests_within_docker]: fixed

File: pkg/inject/secretless.go (hunk 4 above)
Comment (hughsaunders) [reply pkg/inject/secretless.go]: At this point in the code, we're creating a container from scratch, so the mount can't exist. If a duplicate mount get's created, it will be by an external process after this. I'm currently testing deliberately injecting a duplicate volume mount to see if it breaks tests.

Comment (hughsaunders) [reply pkg/inject/secretless.go]: Creating duplicate SATVMs does cause failure. However I'm really not sure what to do about it. As pointed out previously we're creating a container from scratch so the volume mount will not already exist for us to check for it. It may be added in a future step/plugin, and hopefully that process will be idempotent. 

On that note [KEP-36](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#mutating-plugin-ordering) Does say that mutating webhooks that get re-invoked will have to be idempotent. So any plugin that sees the pod spec after injection must be idempotent so it shouldn't add the SATVM if it's already there.

Comment (doodlesbykumbi) [reply pkg/inject/secretless.go]: You're absolutely right. Point 3 in my comment above doesn't make sense since we have no idea if the volume mount will be added after us or not

File: tests/echoserver.yaml.sh [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+  if [[ -z "${ns}" ]]; then
+    echo "default"
+  else
+    echo "${ns}"
+  fi
+}
+
+cat << EOL
+kind: ClusterRole
```
Comment (doodlesbykumbi) [inline tests/echoserver.yaml.sh]: We can use a ClusterRole here to isolate RBAC to the test namespaces. This will help avoid flaky tests in case someone or a competing test might delete the ClusterRole
